### PR TITLE
fix: improve populate type narrowing 

### PR DESCRIPTION
**Pull request by mohammad0-0ahmad**
Should resolve https://github.com/Automattic/mongoose/issues/11496
@Uzlopak

### CODE REVIEW

**Comment by Uzlopak** [comment]
This one is better.

File: types/index.d.ts
Comment (vkarpov15) [inline types/index.d.ts:1816]: Shouldn't the last bit be something like `T[P] & U[P]`? I think `T` should come first, and then we should be merging T's keys as well as opposed to merging T into U's keys.

File: test/types/populate.test.ts
Comment (vkarpov15) [inline test/types/populate.test.ts:179]: Can we do `expectType<boolean>` here?

Comment (mohammad0-0ahmad) [reply test/types/populate.test.ts:179]: I've tried to write two tests for this & it doesn't success, although it shows correct intellisens even in map function.

Untitled

File: types/index.d.ts
Comment (mohammad0-0ahmad) [reply types/index.d.ts:1816]: Check the new commit

File: test/types/populate.test.ts
Comment (Uzlopak) [reply test/types/populate.test.ts]: So is it now fixed?

File: types/index.d.ts
Comment (Uzlopak) [reply types/index.d.ts:1820]: Well hmm, actually the thing is, that it is only a HydratedDocument if in populate the option lean is set to false or lean() is used on the query builder. So I suspect this is wrong :/

Comment (mohammad0-0ahmad) [reply types/index.d.ts:1820]: Alright, I will refact that in the next commit :)

Comment (Uzlopak) [reply types/index.d.ts:1820]: @mohammad0-0ahmad 
Imho, if you cant fix that than I would actually recommend to remove those expectTypes, which vkarpov15 requested and revert those HydratedDocument changes, if this is a hard case to solve for you. (Not suggesting that this is over your skills). Probably this is happening because tsd uses a different typescript version and thus resulting in some issues with latest typescript, which is used in vscode for their intellisense. So this is maybe some kind of scope creep. It was not even originally a task for you. So I actually dont want that you feel forced to solve this issue.

But if you think this is a nice challenge and you really want to supply an A-Grade solution, than feel free to invest more time and energy. And know that your work is atleast by me appreciated.

**Comment by mohammad0-0ahmad** [comment]
I've refacted this & it should work with both VSC intellisense & tsd.

File: types/index.d.ts
Comment (mohammad0-0ahmad) [reply types/index.d.ts]: Thanks for your words, no worries it was funny actually  at the beginning I though that as well but it's works now.

File: types/index.d.ts
Comment (Uzlopak) [inline types/index.d.ts:2122]: In #11518 we got a regression reported. 

```suggestion
    populate<Paths = {}>(options: string[] | PopulateOptions | Array<PopulateOptions>): QueryWithHelpers<UnpackedIntersection<ResultType, Paths>, DocType, THelpers, RawDocType>;
```

Corresponding typing test would be:


```typescript
async function gh11518() {
  // populate should accept an Array with the fields 
  await Story.findOne().populate(["author"]);
}
```

If you want, we can do this in this PR. If not, I can create a PR. I just didnt wanted to create a PR before this is merged to avoid high probable merge conflict.

Comment (Uzlopak) [reply types/index.d.ts:2122]: @vkarpov15 Only the second populate type can handle string arrays correct?

Comment (mohammad0-0ahmad) [reply types/index.d.ts:2122]: OR maybe only the first one?

Comment (Uzlopak) [reply types/index.d.ts:2122]: lets wait for feedback by vkarpov

Comment (vkarpov15) [reply types/index.d.ts:2122]: That's an independent issue that I fixed in #11518, no need to discuss that in this PR. But yes, you can call both syntaxes with an array of paths.

File: types/index.d.ts
Comment (vkarpov15) [reply types/index.d.ts:1817]: We introduced `UnpackedIntersectionWithNull` to fix #11041. Although it looks like the test for that issue still passes with this change, so looks like we don't need it after all.

**Review by vkarpov15 — APPROVED** [review]
Thanks!

**Comment by mohammad0-0ahmad** [comment]
> Thanks!

I've to resolve conflicts

**Pull request merged**
